src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -387,7 +486,87 @@ internal void Run<TState>(string input, int startat, ref TState state, MatchCall
+                while (true)
+                {
+                    runner.InitializeForScan(this, input, startat, false);
+                    runner.runtext = input;
+                    runner.runtextpos = runtextpos;
+
+                    int stoppos = RightToLeft ? 0 : input.Length;
+
+                    Match? match = InternalPerformScan(reuseMatchObject, input, 0, runner, input, returnNullIfQuick: false);
+
+                    // if we got a match, then call the callback function with the match and prepare for next iteration.
+                    if (match!.Success)
+                    {
+                        if (!reuseMatchObject)
+                        {
+                            // We're not reusing match objects, so null out our field reference to the instance.
+                            // It'll be recreated the next time one is needed.
+                            runner.runmatch = null;
+                        }
+
+                        if (!callback(ref state, match))
+                        {
+                            // If the callback returns false, we're done.
+                            // Drop reference to text to avoid keeping it alive in a cache.
+                            runner.runtext = null!;
+
+                            if (reuseMatchObject)
+                            {
+                                // We're reusing the single match instance, so clear out its text as well.
+                                // We don't do this if we're not reusing instances, as in that case we're
+                                // dropping the whole reference to the match, and we no longer own the instance
+                                // having handed it out to the callback.
+                                match.Text = null!;
+                            }
+                            return;
+                        }
+
+                        // Now that we've matched successfully, update the starting position to reflect
+                        // the current position, just as Match.NextMatch() would pass in _textpos as textstart.
+                        runtextpos = startat = runner.runtextpos;
+
+                        // Reset state for another iteration.
+                        runner.runtrackpos = runner.runtrack!.Length;
+                        runner.runstackpos = runner.runstack!.Length;
+                        runner.runcrawlpos = runner.runcrawl!.Length;
+
+                        if (match.Length == 0)
+                        {
+                            if (runner.runtextpos == stoppos)
+                            {
+                                // Drop reference to text to avoid keeping it alive in a cache.
+                                runner.runtext = null!;
+
+                                if (reuseMatchObject)
+                                {
+                                    // See above comment.
+                                    match.Text = null!;
```

Comment:
You changed Text to be nullable, right?  In which case:
```suggestion
                                    match.Text = null;
```